File: apps/api/app/llm/deterministic.py
"""Deterministic fallback logic (ported from apps/web/lib/triage/deterministic.ts)."""

from __future__ import annotations

from datetime import datetime, timedelta
from typing import Literal

from ..models import HealthAnomalyAlert, PatientDecision, TriageRequest, TriageOutcome, TimeSlot, BookingConfirmation


def determine_urgency(
    anomaly_score: int,
    flags: list[str],
    thresholds: dict | None = None,
) -> Literal["routine", "soon", "urgent"]:
    urgent_threshold = (thresholds or {}).get("urgent", 85)
    soon_threshold = (thresholds or {}).get("soon", 70)
    if anomaly_score >= urgent_threshold:
        return "urgent"
    if "SLEEP_DROP" in flags and "RHR_SPIKE" in flags:
        return "urgent"
    if anomaly_score >= soon_threshold:
        return "soon"
    return "routine"


def should_contact_clinic(urgency: str) -> bool:
    return urgency != "routine"


def generate_questions(urgency: str, flags: list[str]) -> list[str]:
    questions: list[str] = []
    if "SLEEP_DROP" in flags:
        questions.append("Have you had trouble falling or staying asleep recently?")
        questions.append("Have you changed your bedtime routine or started any new medications?")
    if "RHR_SPIKE" in flags:
        questions.append("Have you experienced heart palpitations, chest tightness, or shortness of breath?")
        questions.append("Have you been under unusual stress or consumed more caffeine than normal?")
    if "STEPS_DROP" in flags:
        questions.append("Have you been less physically active due to pain, fatigue, or other symptoms?")
    if "HRV_DROP" in flags:
        questions.append("Have you been feeling more anxious or fatigued than usual?")

    generic = [
        "Are you currently experiencing any pain or discomfort?",
        "Have you noticed any other changes in how you feel day-to-day?",
        "Are you taking all prescribed medications as directed?",
        "Have you had any recent illness, injury, or major life changes?",
    ]
    while len(questions) < 3:
        questions.append(generic[len(questions)])

    return questions[:6]


def generate_patient_decision(anomaly: HealthAnomalyAlert) -> PatientDecision:
    urgency = determine_urgency(anomaly.anomaly_score, anomaly.flags)
    contact_clinic = should_contact_clinic(urgency)
    questions = generate_questions(urgency, anomaly.flags)

    flag_desc = (
        " and ".join(f.replace("_", " ").lower() for f in anomaly.flags)
        if anomaly.flags
        else "minor deviations"
    )

    suffix = (
        " We recommend contacting your clinic."
        if urgency != "routine"
        else " No immediate action needed, but keep monitoring."
    )

    return PatientDecision(
        summary_explanation=(
            f"Your wearable data from {anomaly.date} shows {flag_desc} compared to your 28-day baseline. "
            f"Your anomaly score is {anomaly.anomaly_score}/100, which is classified as \"{urgency}\".{suffix}"
        ),
        questions=questions,
        recommended_next_step=(
            f"Schedule a {'same-day' if urgency == 'urgent' else 'follow-up'} appointment with Dr. Smith."
            if contact_clinic
            else "Continue monitoring your metrics. Re-check in 48 hours."
        ),
        should_contact_clinic=contact_clinic,
        urgency=urgency,
        clinic_message=(
            f"Patient {anomaly.user_handle} health anomaly: score {anomaly.anomaly_score}/100, "
            f"flags: {', '.join(anomaly.flags)}. Urgency: {urgency}. "
            f"Please schedule {'immediate' if urgency == 'urgent' else 'a follow-up'} appointment."
            if contact_clinic
            else None
        ),
    )


def generate_mock_slots(urgency: str) -> list[TimeSlot]:
    now = datetime.now()
    slots: list[TimeSlot] = []

    if urgency == "urgent":
        day_offset = 0
    elif urgency == "soon":
        day_offset = 1
    else:
        day_offset = 3

    hours = [9, 11, 14] if urgency == "urgent" else [10, 13, 15]

    for i in range(3):
        d = now + timedelta(days=day_offset + i)
        # Skip weekends
        while d.weekday() >= 5:
            d += timedelta(days=1)

        d = d.replace(hour=hours[i], minute=0, second=0, microsecond=0)
        end = d.replace(minute=30)
        slots.append(TimeSlot(start=d.isoformat(), end=end.isoformat()))

    return slots


def generate_triage_outcome(req: TriageRequest) -> TriageOutcome:
    questions = [
        "How long have you been experiencing these symptoms?",
        "On a scale of 1-10, how would you rate your discomfort right now?",
        "Do you have any known allergies or chronic conditions?",
    ]

    default_answers = {
        questions[0]: "A few days",
        questions[1]: "7" if req.urgency == "urgent" else ("5" if req.urgency == "soon" else "3"),
        questions[2]: "No known allergies",
    }

    answers = (
        req.patient_answers
        if req.patient_answers and len(req.patient_answers) > 0
        else default_answers
    )

    slots = generate_mock_slots(req.urgency)
    chosen_slot = slots[0]
    method: Literal["in_person", "telehealth"] = "in_person" if req.urgency == "urgent" else "telehealth"

    return TriageOutcome(
        intake_questions_asked=questions,
        intake_answers=answers,
        urgency=req.urgency,
        proposed_slots=slots,
        booking_confirmation=BookingConfirmation(
            start=chosen_slot.start,
            end=chosen_slot.end,
            method=method,
        ),
        escalation_triggered=(req.urgency == "urgent" and "RHR_SPIKE" in req.anomaly.flags),
    )

File: apps/web/lib/triage/deterministic.ts
import type { HealthAnomalyAlert, PatientDecision, TriageRequest, TriageOutcome } from "@people/shared";

export function determineUrgency(
  anomalyScore: number,
  flags: string[],
  _userContext?: string
): "routine" | "soon" | "urgent" {
  if (anomalyScore >= 85) return "urgent";
  if (flags.includes("SLEEP_DROP") && flags.includes("RHR_SPIKE")) return "urgent";
  if (anomalyScore >= 70) return "soon";
  return "routine";
}

export function shouldContactClinic(urgency: "routine" | "soon" | "urgent"): boolean {
  return urgency !== "routine";
}

export function generateQuestions(
  urgency: "routine" | "soon" | "urgent",
  flags: string[]
): string[] {
  const questions: string[] = [];

  if (flags.includes("SLEEP_DROP")) {
    questions.push("Have you had trouble falling or staying asleep recently?");
    questions.push("Have you changed your bedtime routine or started any new medications?");
  }
  if (flags.includes("RHR_SPIKE")) {
    questions.push("Have you experienced heart palpitations, chest tightness, or shortness of breath?");
    questions.push("Have you been under unusual stress or consumed more caffeine than normal?");
  }
  if (flags.includes("STEPS_DROP")) {
    questions.push("Have you been less physically active due to pain, fatigue, or other symptoms?");
  }
  if (flags.includes("HRV_DROP")) {
    questions.push("Have you been feeling more anxious or fatigued than usual?");
  }

  // Pad to at least 3
  const generic = [
    "Are you currently experiencing any pain or discomfort?",
    "Have you noticed any other changes in how you feel day-to-day?",
    "Are you taking all prescribed medications as directed?",
    "Have you had any recent illness, injury, or major life changes?",
  ];
  while (questions.length < 3) {
    questions.push(generic[questions.length]!);
  }

  return questions.slice(0, 6);
}

export function generatePatientDecision(anomaly: HealthAnomalyAlert): PatientDecision {
  const urgency = determineUrgency(anomaly.anomaly_score, anomaly.flags, anomaly.freeform_context);
  const contactClinic = shouldContactClinic(urgency);
  const questions = generateQuestions(urgency, anomaly.flags);

  const flagDesc = anomaly.flags.length > 0
    ? anomaly.flags.map(f => f.replace(/_/g, " ").toLowerCase()).join(" and ")
    : "minor deviations";

  return {
    summary_explanation: `Your wearable data from ${anomaly.date} shows ${flagDesc} compared to your 28-day baseline. Your anomaly score is ${anomaly.anomaly_score}/100, which is classified as "${urgency}".${urgency !== "routine" ? " We recommend contacting your clinic." : " No immediate action needed, but keep monitoring."}`,
    questions,
    recommended_next_step: contactClinic
      ? `Schedule a ${urgency === "urgent" ? "same-day" : "follow-up"} appointment with Dr. Smith.`
      : "Continue monitoring your metrics. Re-check in 48 hours.",
    should_contact_clinic: contactClinic,
    urgency,
    clinic_message: contactClinic
      ? `Patient ${anomaly.user_handle} health anomaly: score ${anomaly.anomaly_score}/100, flags: ${anomaly.flags.join(", ")}. Urgency: ${urgency}. Please schedule ${urgency === "urgent" ? "immediate" : "a follow-up"} appointment.`
      : undefined,
  };
}

export function generateTriageOutcome(req: TriageRequest): TriageOutcome {
  const questions = [
    "How long have you been experiencing these symptoms?",
    "On a scale of 1-10, how would you rate your discomfort right now?",
    "Do you have any known allergies or chronic conditions?",
  ];

  const defaultAnswers: Record<string, string> = {
    [questions[0]]: "A few days",
    [questions[1]]: req.urgency === "urgent" ? "7" : req.urgency === "soon" ? "5" : "3",
    [questions[2]]: "No known allergies",
  };

  const answers = req.patient_answers && Object.keys(req.patient_answers).length > 0
    ? req.patient_answers
    : defaultAnswers;

  const slots = generateMockSlots(req.urgency);
  const chosenSlot = slots[0];
  const method = req.urgency === "urgent" ? "in_person" as const : "telehealth" as const;

  return {
    intake_questions_asked: questions,
    intake_answers: answers,
    urgency: req.urgency,
    proposed_slots: slots,
    booking_confirmation: {
      start: chosenSlot.start,
      end: chosenSlot.end,
      method,
    },
    escalation_triggered: req.urgency === "urgent" && req.anomaly.flags.includes("RHR_SPIKE"),
  };
}

function generateMockSlots(urgency: "routine" | "soon" | "urgent"): { start: string; end: string }[] {
  const now = new Date();
  const slots: { start: string; end: string }[] = [];

  let dayOffset: number;
  if (urgency === "urgent") dayOffset = 0;
  else if (urgency === "soon") dayOffset = 1;
  else dayOffset = 3;

  for (let i = 0; i < 3; i++) {
    const d = new Date(now);
    d.setDate(d.getDate() + dayOffset + i);
    // Skip weekends
    while (d.getDay() === 0 || d.getDay() === 6) {
      d.setDate(d.getDate() + 1);
    }

    const hours = urgency === "urgent" ? [9, 11, 14] : [10, 13, 15];
    d.setHours(hours[i], 0, 0, 0);
    const end = new Date(d);
    end.setMinutes(30);

    slots.push({
      start: d.toISOString(),
      end: end.toISOString(),
    });
  }

  return slots;
}
